04-cs-fundamentals/internship/binary-search-trees.md

```diff
@@ -1,23 +1,253 @@
-+ Revise or author exercises related to recursion. See [Assignments and Additional Resources for Recursion](https://github.com/Ada-Developers-Academy/textbook-curriculum/blob/master/04-cs-fundamentals/classroom/Introduction%20to%20Recursion.md#assignments)
-+ Before going through iterative approaches to BST algorithms, revise [Stacks and Queues](https://github.com/Ada-Developers-Academy/textbook-curriculum/blob/master/04-cs-fundamentals/internship/Stacks%20and%20Queues.md)
 
+By the end of this lesson you should be able to:
+
+- Compare a Binary Tree to a Linked List
+- Explain how a Binary Search Tree differs from a generic Binary Tree
+- Write methods to perform the following on a Binary Search tree:
+  - Search
+  - Insert value
+  - Delete value
+  - Find height
+  - Perform traversals including: 
+    - Depth first traversals: pre-order, in-order, post-order
+    - Breadth first traversal
+
+## Video Lesson
+
+- [Video Lesson](https://adaacademy.hosted.panopto.com/Panopto/Pages/Viewer.aspx?id=d9746397-8a10-43be-b1cc-aaaf00720b31)
+- [Slide Deck](https://docs.google.com/presentation/d/1jE2bQDXkoCIJmqBgFXuripXftzrijuSY1JRKf2Ln7MY/edit#slide=id.p)
+
+
+## Limitations of Linked Lists
+
+Remember that finding an element in a Linked List requires a runtime of O(n).  This is because any search in a Linked List requires a linear search.  However once a Linked List node is found, removing or adding a node is O(1).  It would be ideal to combine the binary search O(log<sub>n</sub>) or a sorted array with the O(1) runtime to remove or add nodes.  This is what Binary Search Trees give us.
+
+Instead of a node with one link to a next node, we can create nodes with 2 pointers, left and right.  Each node is not allowed to point back to a parent or grandparent.   So you can arrive at a child by following the links of it's parents, but there is no path from a child to a parent.  This ends up forming a tree like shape with each node having 0, 1 or 2 "children".  This is called a _Binary Tree_.  In this way a binary tree is an _acyclic, connected graph_.  It forms no cycles or loops of links, and each node is connected to others in the data structure.
+
+In a special type of Binary Tree, called a _Binary Search Tree_ the left pointer points to all elements smaller than the node.  The right pointer points to all nodes greater than the given node.   Each node can refer to other nodes, but there is no way to go from a child node to a parent.  The links go in one direction.  A Tree is hierarchical with certain nodes acting as parents to others.  
+
+![bst.png](images/bst.png)
+
+Notice that nodes 61 and 77, both form subtrees of the larger tree data structure.  This enables you to use **recursion** on most tree methods because each child node forms a subtree.
+
+**Exercise**
+
+Given the following values draw a binary search tree:
+
+"Ringo", "John", "Paul", "George"
```

Comment:
This is tricky because there are many valid ways to organize these into a BST, depending on insertion order. That ambiguity makes me a little uncomfortable for an exercise prompt.

If the goal of this question is to reinforce "what is a binary tree / what is a BST", maybe you could do something like providing a few images, and asking students to identify which ones are binary trees, which are BSTs, and which are neither.